2.0.0 :small:`March 14, 2022`
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Overall:**

* Cumulus workflows are now released on Dockstore_:

    * Add the tutorial on `importing Cumulus workflows to Terra`_.
    * Archive the `legacy versions`_ on Broad Method Registry.

* Add support on multiple platforms via **backend** input: ``gcp`` for Google Cloud, ``aws`` for Amazon AWS, ``local`` for local machine. Enable Google Cloud support by default.

* For Amazon AWS backend, add **awsMaxRetries** input to set the maximum retries allowed for job execution at runtime. By default, use ``5``.

* Update the `command-line job submission`_ tutorial to work with Altocumulus_ v2.0.0 or later.

* On Examples:

    * Update `gene expression, hashing and CITE-Seq`_ example tutorial.
    * Add tutorial on `10x CellPlex analysis`_ using Cumulus workflows on Cloud.

**Workflow-specific:**

* Add *STARsolo_create_reference* workflow to build genome references for STARsolo counting. See its `documentation <./starsolo.html#prepare-a-sample-sheet>`_ for details.

* On *Cellranger* workflow:

    * Add support for 10x Cell Ranger version ``6.1.1`` and ``6.1.2``, and use ``6.1.2`` by default. See `Cell Ranger v6.1 release notes`_.
    * Add support for 10x Cell Ranger ARC version ``2.0.1``, and use it by default. See `Cell Ranger ARC v2.0 release notes`_ for the release notes.
    * Upgrade cumulus_feature_barcoding_ to version ``0.7.0`` to allow manually set barcode starting position (via input **crispr_barcode_pos**).
    * Add support for non 10x CRISPR assays. See the description of ``crispr`` **DataType** value in `this section <./cellranger/index.html#id1>`_ for details.
    * For input data consisting of fastq files, it's able to handle folder structure of both flat (all fastq files in one folder) and nested (one subfolder per sample listed in the input sample sheet) forms.
    * Add **fastq_outputs** to workflow output, which contains *mkfastq* step output folders for samples listed in the input sample sheet.
    * Add **count_outputs** to workflow output, which contains *count* step output folderrs for samples listed in the input sample sheet.

* On *Spaceranger* workflow:

    * Add support for 10x Space Ranger version ``1.3.0`` and ``1.3.1``, and use ``1.3.1`` by default. See `Space Ranger v1.3 release notes`_ for the release notes.
    * For input data consisting of fastq files, it's able to handle folder structure of both flat (all fastq files in one folder) and nested (one subfolder per library) forms.
    * Add output section for the workflow. See `here <./spaceranger.html#workflow-output>`_ for details.
    * Retire old genome references:

        * Keep ``GRCh38-2020-A`` and ``mm10-2020-A``.
        * Retire ``GRCh38``, ``mm10``, ``GRCh38-2020-A-premrna`` and ``mm10-2020-A-premrna``. Users can still reach out to `Cumulus team`_ to ask for URIs to these old references, but they are not provided by default.
    * In the description of **ReorientImages** field of input sample sheet, add the information on its valid values.

* On *STARsolo* workflow:

    * Add support for STAR version ``2.7.9a``, and use it by default. See `STAR v2.7.9a release notes`_ for the release notes.
    * Reorganize the workflow by exposing more inputs to users.
    * Add support on more protocols: 10x multiome, 10x 5' (both SC5P-R2 and SC5P-PE), Slide-Seq and Share-Seq. See `here <./starsolo.html#prepare-a-sample-sheet>` for details.
    * Use input **read1_fastq_pattern** and **read2_fastq_pattern** to support fastq files generated by Cell Ranger or SeqWell, as well as Sequence Read Archive (SRA) data.
    * For input data consisting of fastq files, it's able to handle folder structure of both flat (all fastq files in one folder) and nested (one subfolder per library) forms.
    * Do not attach filename prefix to output files to avoid the incorrect SJ raw *feature.tsv* symlink error, which would cause the folder delocalization fail. (see `discussion <https://github.com/alexdobin/STAR/issues/1458>`_ with STAR_ team)
    * Add STAR log file to workflow output. This is the *Log.out* file if running STAR locally, which can be used for tracking the process and sharing with STAR_ team when opening an issue there.
    * Retire old genome references:

        * Keep ``GRCh38-2020-A``, ``mm10-2020-A``, and ``GRCh38-and-mm10-2020-A``.
        * Retire old references listed `here <https://cumulus.readthedocs.io/en/1.5.0/spaceranger.html#sample-sheet>`_. Users can still reach out to `Cumulus team`_ to ask for URIs to them, but they are not provided by default.

* On *Demultiplexing* workflow:

    * Upgrade demuxEM_ to version ``0.1.7`` for bug fix.

* On *Cellranger_create_reference* workflow:

    * Add the generated reference file to the workflow output.
    * Bug fix in using input **memory**.
    * Update documentation to suggest only using Cell Ranger version ``6.1.1`` or later for building reference, as v6.0.1 has issues which leave the job running without terminating.

* On *Cellranger_atac_create_reference* workflow:

    * Add the generated reference file to the workflow output.

* On *Cellranger_vdj_create_reference* workflow:

    * Add the generated reference file to the workflow output.

.. _Dockstore: https://dockstore.org/organizations/lilab/collections/Cumulus
.. _importing Cumulus workflows to Terra: ./cumulus_import.html
.. _legacy versions: ./workflow_overview/broad_method_registry.html
.. _cumulus_feature_barcoding: https://github.com/lilab-bcb/cumulus_feature_barcoding
.. _gene expression, hashing and CITE-Seq: ./examples/example_hashing_citeseq.html
.. _10x CellPlex analysis: ./examples/example_cellplex.html
.. _demuxEM: https://demuxem.readthedocs.io
.. _command-line job submission: ./command_line.html
.. _Altocumulus: https://altocumulus.readthedocs.io
.. _STAR: https://github.com/alexdobin/STAR
.. _Cumulus team: ./contact.html
.. _Cell Ranger v6.1 release notes: https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/6.1/release-notes
.. _Cell Ranger ARC v2.0 release notes: https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/2.0/release-notes
.. _Space Ranger v1.3 release notes: https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/1.3/release-notes
.. _STAR v2.7.9a release notes: https://github.com/alexdobin/STAR/releases/tag/2.7.9a
