To process sc/snRNA-seq data, follow the specific instructions below.

Sample sheet
++++++++++++

#. **Reference** column.

	Pre-built scRNA-seq references are summarized below.

	.. list-table::
		:widths: 5 20
		:header-rows: 1

		* - Keyword
		  - Description
		* - **GRCh38-2020-A**
		  - Human GRCh38 (GENCODE v32/Ensembl 98)
		* - **mm10-2020-A**
		  - Mouse mm10 (GENCODE vM23/Ensembl 98)
		* - **GRCh38_and_mm10-2020-A**
		  - Human GRCh38 (GENCODE v32/Ensembl 98) and mouse mm10 (GENCODE vM23/Ensembl 98)
		* - **GRCh38_v3.0.0**
		  - Human GRCh38, cellranger reference 3.0.0, Ensembl v93 gene annotation
		* - **hg19_v3.0.0**
		  - Human hg19, cellranger reference 3.0.0, Ensembl v87 gene annotation
		* - **mm10_v3.0.0**
		  - Mouse mm10, cellranger reference 3.0.0, Ensembl v93 gene annotation
		* - **GRCh38_and_mm10_v3.1.0**
		  - Human (GRCh38) and mouse (mm10), cellranger references 3.1.0, Ensembl v93 gene annotations for both human and mouse
		* - **hg19_and_mm10_v3.0.0**
		  - Human (hg19) and mouse (mm10), cellranger reference 3.0.0, Ensembl v93 gene annotations for both human and mouse
		* - **GRCh38_v1.2.0** or **GRCh38**
		  - Human GRCh38, cellranger reference 1.2.0, Ensembl v84 gene annotation
		* - **hg19_v1.2.0** or **hg19**
		  - Human hg19, cellranger reference 1.2.0, Ensembl v82 gene annotation
		* - **mm10_v1.2.0** or **mm10**
		  - Mouse mm10, cellranger reference 1.2.0, Ensembl v84 gene annotation
		* - **GRCh38_and_mm10_v1.2.0** or **GRCh38_and_mm10**
		  - Human and mouse, built from GRCh38 and mm10 cellranger references, Ensembl v84 gene annotations are used
		* - **GRCh38_and_SARSCoV2**
		  - Human GRCh38 and SARS-COV-2 RNA genome, cellranger reference 3.0.0, generated by `Carly Ziegler`_. The SARS-COV-2 viral sequence and gtf are as described in `[Kim et al. Cell 2020]`_ (https://github.com/hyeshik/sars-cov-2-transcriptome, BetaCov/South Korea/KCDC03/2020 based on NC_045512.2). The GTF was edited to include only CDS regions, and regions were added to describe the 5' UTR ("SARSCoV2_5prime"), the 3' UTR ("SARSCoV2_3prime"), and reads aligning to anywhere within the Negative Strand("SARSCoV2_NegStrand"). Additionally, trailing A's at the 3' end of the virus were excluded from the SARSCoV2 fasta, as these were found to drive spurious viral alignment in pre-COVID19 samples.

	Pre-built snRNA-seq references are summarized below.

	.. list-table::
		:widths: 5 20
		:header-rows: 1

		* - Keyword
		  - Description
		* - **GRCh38_premrna_v3.0.0**
		  - Human, introns included, built from GRCh38 cellranger reference 3.0.0, Ensembl v93 gene annotation, treating annotated transcripts as exons
		* - **GRCh38_premrna_v1.2.0** or **GRCh38_premrna**
		  - Human, introns included, built from GRCh38 cellranger reference 1.2.0, Ensembl v84 gene annotation, treating annotated transcripts as exons
		* - **mm10_premrna_v1.2.0** or **mm10_premrna**
		  - Mouse, introns included, built from mm10 cellranger reference 1.2.0, Ensembl v84 gene annotation, treating annotated transcripts as exons
		* - **GRCh38_premrna_and_mm10_premrna_v1.2.0** or **GRCh38_premrna_and_mm10_premrna**
		  - Human and mouse, introns included, built from GRCh38_premrna_v1.2.0 and mm10_premrna_v1.2.0
		* - **GRCh38_premrna_and_SARSCoV2**
		  - Human, introns included, built from GRCh38_premrna_v3.0.0, and SARS-COV-2 RNA genome. This reference was generated by `Carly Ziegler`_. The SARS-COV-2 RNA genome is from `[Kim et al. Cell 2020]`_ (https://github.com/hyeshik/sars-cov-2-transcriptome, BetaCov/South Korea/KCDC03/2020 based on NC_045512.2). Please see the description of *GRCh38_and_SARSCoV2* above for details.

#. **Index** column.

	Put `10x single cell RNA-seq sample index set names`_ (e.g. SI-GA-A12) here.

#. *Chemistry* column.

	According to *cellranger count*'s documentation, chemistry can be

	.. list-table::
		:widths: 5 20
		:header-rows: 1

		* - Chemistry
		  - Explanation
		* - **auto**
		  - autodetection (default). If the index read has extra bases besides cell barcode and UMI, autodetection might fail. In this case, please specify the chemistry
		* - **threeprime**
		  - Single Cell 3′
		* - **fiveprime**
		  - Single Cell 5′
		* - **SC3Pv1**
		  - Single Cell 3′ v1
		* - **SC3Pv2**
		  - Single Cell 3′ v2
		* - **SC3Pv3**
		  - Single Cell 3′ v3. You should set cellranger version input parameter to >= 3.0.2
		* - **SC5P-PE**
		  - Single Cell 5′ paired-end (both R1 and R2 are used for alignment)
		* - **SC5P-R2**
		  - Single Cell 5′ R2-only (where only R2 is used for alignment)

#. *DataType* column.

	This column is optional with a default **rna**. If you want to put a value, put **rna** here.

#. *FetureBarcodeFile* column.

	Put target panel CSV file here for targeted expressiond data. Note that if a target panel CSV is present, cell ranger version must be >= 4.0.0.

#. Example::

	Sample,Reference,Flowcell,Lane,Index,Chemistry,DataType,FeatureBarcodeFile
	sample_1,GRCh38-2020-A,gs://fc-e0000000-0000-0000-0000-000000000000/VK18WBC6Z4,1-2,SI-GA-A8,threeprime,rna
	sample_1,GRCh38-2020-A,gs://fc-e0000000-0000-0000-0000-000000000000/VK10WBC9Z2,1-2,SI-GA-A8,threeprime,rna
	sample_2,mm10-2020-A,gs://fc-e0000000-0000-0000-0000-000000000000/VK18WBC6Z4,5-6,SI-GA-C8,fiveprime,rna
	sample_2,mm10-2020-A,gs://fc-e0000000-0000-0000-0000-000000000000/VK10WBC9Z2,5-6,SI-GA-C8,fiveprime,rna
	sample_3,GRCh38-2020-A,gs://fc-e0000000-0000-0000-0000-000000000000/VK18WBC6Z4,3,SI-TT-A1,auto,rna,gs://fc-e0000000-0000-0000-0000-000000000000/immunology_v1.0_GRCh38-2020-A.target_panel.csv

Workflow input
++++++++++++++

For sc/snRNA-seq data, ``cellranger_workflow`` takes Illumina outputs as input and runs ``cellranger mkfastq`` and ``cellranger count``. Revalant workflow inputs are described below, with required inputs highlighted in bold.

	.. list-table::
		:widths: 5 30 30 20
		:header-rows: 1

		* - Name
		  - Description
		  - Example
		  - Default
		* - **input_csv_file**
		  - Sample Sheet (contains Sample, Reference, Flowcell, Lane, Index as required and Chemistry, DataType, FeatureBarcodeFile as optional)
		  - "gs://fc-e0000000-0000-0000-0000-000000000000/sample_sheet.csv"
		  -
		* - **output_directory**
		  - Output directory
		  - "gs://fc-e0000000-0000-0000-0000-000000000000/cellranger_output"
		  - Results are written under directory *output_directory* and will overwrite any existing files at this location.
		* - run_mkfastq
		  - If you want to run ``cellranger mkfastq``
		  - true
		  - true
		* - run_count
		  - If you want to run ``cellranger count``
		  - true
		  - true
		* - delete_input_bcl_directory
		  - If delete BCL directories after demux. If false, you should delete this folder yourself so as to not incur storage charges
		  - false
		  - false
		* - mkfastq_barcode_mismatches
		  - Number of mismatches allowed in matching barcode indices (bcl2fastq2 default is 1)
		  - 0
		  -
		* - mkfastq_filter_single_index
		  - Only demultiplex samples identified by an i7-only sample index, ignoring dual-indexed samples. Dual-indexed samples will not be demultiplexed
		  - false
		  - false
		* - mkfastq_use_bases_mask
		  - Override the read lengths as specified in *RunInfo.xml*
		  - "Y28n*,I8n*,N10,Y90n*"
		  -
		* - mkfastq_delete_undetermined
		  - Delete undetermined FASTQ files generated by bcl2fastq2
		  - true
		  - false
		* - force_cells
		  - Force pipeline to use this number of cells, bypassing the cell detection algorithm, mutually exclusive with expect_cells
		  - 6000
		  -
		* - expect_cells
		  - Expected number of recovered cells. Mutually exclusive with force_cells
		  - 3000
		  -
		* - include_introns
		  - Turn this option on to also count reads mapping to intronic regions. With this option, users do not need to use pre-mRNA references. Note that if this option is set, cellranger_version must be >= 5.0.0.
		  - false
		  - false
		* - no_bam
		  - Turn this option on to disable BAM file generation. This option is only available if cellranger_version >= 5.0.0.
		  - false
		  - false
		* - secondary
		  - Perform Cell Ranger secondary analysis (dimensionality reduction, clustering, etc.)
		  - false
		  - false
		* - cellranger_version
		  - cellranger version, could be 6.1.1, 6.0.2, 6.0.1, 6.0.0, 5.0.1, 5.0.0, 4.0.0, 3.1.0, 3.0.2, or 2.2.0
		  - "6.1.1"
		  - "6.1.1"
		* - config_version
		  - config docker version used for processing sample sheets, could be 0.2, 0.1
		  - "0.2"
		  - "0.2"
		* - docker_registry
		  - Docker registry to use for cellranger_workflow. Options:

		  	- "quay.io/cumulus" for images on Red Hat registry;

		  	- "cumulusprod" for backup images on Docker Hub.
		  - "quay.io/cumulus"
		  - "quay.io/cumulus"
		* - mkfastq_docker_registry
		  - Docker registry to use for ``cellranger mkfastq``.
		    Default is the registry to which only Broad users have access.
		    See :ref:`bcl2fastq-docker` for making your own registry.
		  - "gcr.io/broad-cumulus"
		  - "gcr.io/broad-cumulus"
		* - acronym_file
		  - | The link/path of an index file in TSV format for fetching preset genome references, chemistry whitelists, etc. by their names.
		    | Set an GS URI if *backend* is ``gcp``; an S3 URI for ``aws`` backend; an absolute file path for ``local`` backend.
		  - "s3://xxxx/index.tsv"
		  - "gs://regev-lab/resources/cellranger/index.tsv"
		* - zones
		  - Google cloud zones
		  - "us-central1-a us-west1-a"
		  - "us-central1-a us-central1-b us-central1-c us-central1-f us-east1-b us-east1-c us-east1-d us-west1-a us-west1-b us-west1-c"
		* - num_cpu
		  - Number of cpus to request for one node for cellranger mkfastq and cellranger count
		  - 32
		  - 32
		* - memory
		  - Memory size string for cellranger mkfastq and cellranger count
		  - "120G"
		  - "120G"
		* - mkfastq_disk_space
		  - Optional disk space in GB for mkfastq
		  - 1500
		  - 1500
		* - count_disk_space
		  - Disk space in GB needed for cellranger count
		  - 500
		  - 500
		* - backend
		  - Cloud backend for file transfer. Available options:

		    - "gcp" for Google Cloud;
		    - "aws" for Amazon AWS;
		    - "local" for local machine.
		  - "gcp"
		  - "gcp"
		* - preemptible
		  - Number of preemptible tries
		  - 2
		  - 2
		* - awsMaxRetries
		  - Number of maximum retries when running on AWS. This works only when *backend* is ``aws``.
		  - 5
		  - 5

Workflow output
+++++++++++++++

See the table below for important sc/snRNA-seq outputs.

.. list-table::
	:widths: 5 5 10
	:header-rows: 1

	* - Name
	  - Type
	  - Description
	* - cellranger_mkfastq.output_fastqs_directory
	  - Array[String]?
	  - Subworkflow output. A list of cloud urls containing FASTQ files, one url per flowcell.
	* - cellranger_count.output_count_directory
	  - Array[String]?
	  - Subworkflow output. A list of cloud urls containing gene count matrices, one url per sample.
	* - cellranger_count.output_web_summary
	  - Array[File]?
	  - Subworkflow output. A list of htmls visualizing QCs for each sample (cellranger count output).
	* - collect_summaries.metrics_summaries
	  - File?
	  - Task output. A excel spreadsheet containing QCs for each sample.
	* - count_matrix
	  - String
	  - Workflow output. Cloud url for a template count_matrix.csv to run Cumulus.


.. _Carly Ziegler: http://shaleklab.com/author/carly/
.. _[Kim et al. Cell 2020]: https://www.sciencedirect.com/science/article/pii/S0092867420304062
.. _10x single cell RNA-seq sample index set names: https://support.10xgenomics.com/single-cell-gene-expression/index/doc/specifications-sample-index-sets-for-single-cell-3
