To process single-cell immune profiling (scIR-seq) data, follow the specific instructions below.

Sample sheet
++++++++++++

#. **Reference** column.

	Pre-built scIR-seq references are summarized below.

	.. list-table::
		:widths: 5 20
		:header-rows: 1

		* - Keyword
		  - Description
		* - **GRCh38_vdj_v7.0.0**
		  - Human GRCh38 V(D)J sequences, cellranger reference 7.0.0, annotation built from Ensembl *Homo_sapiens.GRCh38.94.chr_patch_hapl_scaff.gtf*
		* - **GRCm38_vdj_v7.0.0**
		  - Mouse GRCm38 V(D)J sequences, cellranger reference 7.0.0, annotation built from Ensembl *Mus_musculus.GRCm38.94.gtf*
		* - **GRCh38_vdj_v5.0.0**
		  - Human GRCh38 V(D)J sequences, cellranger reference 5.0.0, annotation built from Ensembl *Homo_sapiens.GRCh38.94.chr_patch_hapl_scaff.gtf*
		* - **GRCm38_vdj_v5.0.0**
		  - Mouse GRCm38 V(D)J sequences, cellranger reference 5.0.0, annotation built from Ensembl *Mus_musculus.GRCm38.94.gtf*
		* - **GRCh38_vdj_v4.0.0**
		  - Human GRCh38 V(D)J sequences, cellranger reference 4.0.0, annotation built from Ensembl *Homo_sapiens.GRCh38.94.chr_patch_hapl_scaff.gtf*
		* - **GRCm38_vdj_v4.0.0**
		  - Mouse GRCm38 V(D)J sequences, cellranger reference 4.0.0, annotation built from Ensembl *Mus_musculus.GRCm38.94.gtf*
		* - **GRCh38_vdj_v3.1.0**
		  - Human GRCh38 V(D)J sequences, cellranger reference 3.1.0, annotation built from Ensembl *Homo_sapiens.GRCh38.94.chr_patch_hapl_scaff.gtf*
		* - **GRCm38_vdj_v3.1.0**
		  - Mouse GRCm38 V(D)J sequences, cellranger reference 3.1.0, annotation built from Ensembl *Mus_musculus.GRCm38.94.gtf*
		* - **GRCh38_vdj_v2.0.0** or **GRCh38_vdj**
		  - Human GRCh38 V(D)J sequences, cellranger reference 2.0.0, annotation built from Ensembl *Homo_sapiens.GRCh38.87.chr_patch_hapl_scaff.gtf* and *vdj_GRCh38_alts_ensembl_10x_genes-2.0.0.gtf*
		* - **GRCm38_vdj_v2.2.0** or **GRCm38_vdj**
		  - Mouse GRCm38 V(D)J sequences, cellranger reference 2.2.0, annotation built from Ensembl *Mus_musculus.GRCm38.90.chr_patch_hapl_scaff.gtf*

#. **Index** column.

	Put `10x single cell V(D)J sample index set names`_ (e.g. SI-GA-A3) here.

#. *Chemistry* column.

	This column is not used for scIR-seq data. Put **fiveprime** here as a placeholder if you decide to include the Chemistry column.

#. *DataType* column.

	Set it to **vdj**.

#. *FetureBarcodeFile* column.

	Leave it blank for scIR-seq.

#. Example::

	Sample,Reference,Flowcell,Lane,Index,Chemistry,DataType
	sample_vdj,GRCh38_vdj_v3.1.0,gs://fc-e0000000-0000-0000-0000-000000000000/VK10WBC9ZZ,1,SI-GA-A1,fiveprime,vdj

Workflow input
++++++++++++++

For scIR-seq data, ``cellranger_workflow`` takes Illumina outputs as input and runs ``cellranger mkfastq`` and ``cellranger vdj``. Revalant workflow inputs are described below, with required inputs highlighted in bold.

.. list-table::
	:widths: 5 30 30 20
	:header-rows: 1

	* - Name
	  - Description
	  - Example
	  - Default
	* - **input_csv_file**
	  - Sample Sheet (contains Sample, Reference, Flowcell, Lane, Index as required and Chemistry, DataType, FeatureBarcodeFile as optional)
	  - "gs://fc-e0000000-0000-0000-0000-000000000000/sample_sheet.csv"
	  -
	* - **output_directory**
	  - Output directory
	  - "gs://fc-e0000000-0000-0000-0000-000000000000/cellranger_output"
	  -
	* - run_mkfastq
	  - If you want to run ``cellranger mkfastq``
	  - true
	  - true
	* - run_count
	  - If you want to run ``cellranger vdj``
	  - true
	  - true
	* - delete_input_bcl_directory
	  - If delete BCL directories after demux. If false, you should delete this folder yourself so as to not incur storage charges
	  - false
	  - false
	* - mkfastq_barcode_mismatches
	  - Number of mismatches allowed in matching barcode indices (bcl2fastq2 default is 1)
	  - 0
	  -
	* - mkfastq_force_single_index
	  - If 10x-supplied i7/i5 paired indices are specified, but the flowcell was run with only one sample index, allow the demultiplex to proceed using the i7 half of the sample index pair
	  - false
	  - false
	* - mkfastq_filter_single_index
	  - Only demultiplex samples identified by an i7-only sample index, ignoring dual-indexed samples. Dual-indexed samples will not be demultiplexed
	  - false
	  - false
	* - mkfastq_use_bases_mask
	  - Override the read lengths as specified in *RunInfo.xml*
	  - "Y28n*,I8n*,N10,Y90n*"
	  -
	* - mkfastq_delete_undetermined
	  - Delete undetermined FASTQ files generated by bcl2fastq2
	  - true
	  - false
	* - vdj_denovo
	  - Do not align reads to reference V(D)J sequences before de novo assembly
	  - false
	  - false
	* - vdj_chain
	  - Force the analysis to be carried out for a particular chain type. The accepted values are:

		- "auto" for auto detection based on TR vs IG representation;

		- "TR" for T cell receptors;

		- "IG" for B cell receptors.
	  - "auto"
	  - "auto"
	* - cellranger_version
	  - cellranger version, could be 7.0.1, 7.0.0, 6.1.2, 6.1.1, 6.0.2, 6.0.1, 6.0.0, 5.0.1, 5.0.0
	  - "7.0.1"
	  - "7.0.1"
	* - docker_registry
	  - Docker registry to use for cellranger_workflow. Options:

	  	- "quay.io/cumulus" for images on Red Hat registry;

	  	- "cumulusprod" for backup images on Docker Hub.
	  - "quay.io/cumulus"
	  - "quay.io/cumulus"
	* - mkfastq_docker_registry
	  - Docker registry to use for ``cellranger mkfastq``.
	    Default is the registry to which only Broad users have access.
	    See :ref:`bcl2fastq-docker` for making your own registry.
	  - "gcr.io/broad-cumulus"
	  - "gcr.io/broad-cumulus"
	* - acronym_file
	  - | The link/path of an index file in TSV format for fetching preset genome references, chemistry whitelists, etc. by their names.
	    | Set an GS URI if *backend* is ``gcp``; an S3 URI for ``aws`` backend; an absolute file path for ``local`` backend.
	  - "s3://xxxx/index.tsv"
	  - "gs://regev-lab/resources/cellranger/index.tsv"
	* - zones
	  - Google cloud zones
	  - "us-central1-a us-west1-a"
	  - "us-central1-a us-central1-b us-central1-c us-central1-f us-east1-b us-east1-c us-east1-d us-west1-a us-west1-b us-west1-c"
	* - num_cpu
	  - Number of cpus to request for one node for cellranger mkfastq and cellranger vdj
	  - 32
	  - 32
	* - memory
	  - Memory size string for cellranger mkfastq and cellranger vdj
	  - "120G"
	  - "120G"
	* - mkfastq_disk_space
	  - Optional disk space in GB for mkfastq
	  - 1500
	  - 1500
	* - vdj_disk_space
	  - Disk space in GB needed for cellranger vdj
	  - 500
	  - 500
	* - backend
	  - Cloud backend for file transfer. Available options:

	  	- "gcp" for Google Cloud;
	  	- "aws" for Amazon AWS;
	  	- "local" for local machine.
	  - "gcp"
	  - "gcp"
	* - preemptible
	  - Number of preemptible tries
	  - 2
	  - 2
	* - awsQueueArn
	  - The AWS ARN string of the job queue to be used. This only works for ``aws`` backend.
	  - "arn:aws:batch:us-east-1:xxx:job-queue/priority-gwf"
	  - ""


Workflow output
+++++++++++++++

See the table below for important scIR-seq outputs.

.. list-table::
	:widths: 5 5 10
	:header-rows: 1

	* - Name
	  - Type
	  - Description
	* - cellranger_mkfastq.output_fastqs_directory
	  - Array[String]?
	  - Subworkflow output. A list of cloud urls containing FASTQ files, one url per flowcell.
	* - cellranger_vdj.output_vdj_directory
	  - Array[String]?
	  - Subworkflow output. A list of cloud urls containing vdj results, one url per sample.
	* - cellranger_vdj.output_web_summary
	  - Array[File]?
	  - Subworkflow output. A list of htmls visualizing QCs for each sample (cellranger vdj output).
	* - collect_summaries_vdj.metrics_summaries
	  - File?
	  - Task output. A excel spreadsheet containing QCs for each sample.


.. _10x single cell V(D)J sample index set names: https://www.10xgenomics.com/support/single-cell-immune-profiling/documentation/steps/sequencing/sample-index-sets-for-single-cell-immune-profiling
